### PR TITLE
Add a class to log asynchrously.

### PR DESCRIPTION
**Pull request by padenot**
We can only log from a single thread at the moment, we can add an MPSC ring buffer later I suppose. Also will need squashing before merging.

### CODE REVIEW

**Comment by kinetiknz** [comment]
This will be useful and generally looks fine, but I think we need the MSPC queue or some solution to allow logging from any thread, otherwise this is too much of a foot-gun right now.

**Comment by padenot** [comment]
I'll pick that up tomorrow, we need it to debug a real problem during calls. I'll probably remove the ALOGV statements from this patch set, and manually apply gecko-only patches in mozilla-central.

If this class is used from multiple threads, it will blow up immediately anyways. I haven't had time to do a MPSC queue, but that's still planed.

File: src/cubeb_log.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+
+void cubeb_async_log(const char * fmt, ...)
+{
+  // This is going to copy a 256 bytes array around, which is fine.
+  // We don't want to allocate memory here, because this is made to
+  // be called from a real-time callback.
+  va_list args;
+  va_start(args, fmt);
+  char msg[CUBEB_LOG_MESSAGE_MAX_SIZE];
```
Comment (padenot) [inline src/cubeb_log.cpp]: This is just a stack pointer increment, no big deal.

File: src/cubeb_log.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  vsnprintf(msg, CUBEB_LOG_MESSAGE_MAX_SIZE, fmt, args);
+  cubeb_async_logger::get().push(msg);
```
Comment (padenot) [inline src/cubeb_log.cpp]: You're right, I'll fix this.

**Comment by padenot** [comment]
@kinetiknz, in light of https://github.com/kinetiknz/cubeb/pull/205#issuecomment-285120410, can we reconsider merging this? As said previously, I need it to debug a nasty clock drift issue when doing WebRTC calls.

Or maybe it's better to apply this as a patch in `mozilla-central`, considering I'm also going to have invasive logging inside the resampler anyways. What do you think ?

File: src/cubeb_log.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        while (msg_queue.dequeue(&msg, 1)) {
+          LOGV("%s", msg.get());
+        }
+        std::this_thread::sleep_for(CUBEB_LOG_BATCH_PRINT_INTERVAL_MS);
+      }
+    }).detach();
+  }
+private:
+  cubeb_async_logger()
+    :msg_queue(CUBEB_LOG_MESSAGE_QUEUE_DEPTH)
```
Comment (kinetiknz) [inline src/cubeb_log.cpp]: Space after :

File: src/cubeb_log.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  }
+  /** This is quite a big data structure, but is only instantiated if the
+   * asynchronous logger is used.*/
+  lock_free_queue<cubeb_log_message> msg_queue;
+};
+
+
+void cubeb_async_log(const char * fmt, ...)
+{
+  if (g_log_callback) {
```
Comment (kinetiknz) [inline src/cubeb_log.cpp]: !g_log_callback?

File: src/cubeb_log.h [hunk 5]
```diff
@@ -8,6 +8,8 @@
 #ifndef CUBEB_LOG
 #define CUBEB_LOG
 
+#include "cubeb/cubeb.h"
```
Comment (kinetiknz) [inline src/cubeb_log.h]: Why's this needed?

Comment (padenot) [inline src/cubeb_log.h]: `cubeb_log_level`, `cubeb_log_callback`, and the values of `cubeb_log_level`.

**Comment by padenot** [comment]
This takes care of your comments.

**Comment by kinetiknz** [comment]
Merged in d974ad6b5fb71bae0c0089cbeb54645dfe4ecbc9 with minor fixes.

**Pull request closed without merging**